pc-parse/pcpatr/lcparse.c

```diff
@@ -984,255 +1100,302 @@ if (!act_edge->bFailed)
+		pDag1 = findOrAddPATRAttribute(pDag, nterm->pszName, pData->pPATR);
+		unifyPATRFeatures(pDag1, pUniDag, FALSE, pData->pPATR);
 		/*
-		 *  unify lhs features corresponding to nterm with 0 path
+		 *  unify lhs features with 0 path
 		 */
-		if (unifyPATRFeatures(findOrAddPATRAttribute(pDag,
-				  nterm->pszLhsName,
-				  pData->pPATR),
-				  findOrAddPATRAttribute(pDag, psz0, pData->pPATR),
-				  FALSE, pData->pPATR) == NULL)
-		ok = FALSE;
-		else
 		ok = TRUE;
-		if (  ok &&
-		  (act_edge->eType == PATR_RULE_EDGE) &&
-		  (act_edge->u.r.pRule->iNontermCount == act_edge->u.r.iNext) )
+		if (unifyPATRFeatures(findOrAddPATRAttribute(pDag, nterm->pszLhsName, pData->pPATR),
+			findOrAddPATRAttribute(pDag, psz0, pData->pPATR),
+			FALSE, pData->pPATR) == NULL)
 		{
-		PATRPriorityUnion * pUnion;
-		PATRConstraint *    pConstraint;
-		/*
-		 *  apply each priority union in turn
-		 */
-		for (	pUnion = act_edge->u.r.pRule->pPriorityUnions ;
-			ok && pUnion ;
-			pUnion = pUnion->pNext )
-			{
-			applyPATRPriorityUnion(pDag, pUnion, pData->pPATR);
-			ok = verifyAcyclicPATRFeature(pDag);
-			}
-		/*
-		 *  if any constraint fails, ok = FALSE
-		 */
-		for (   pConstraint = act_edge->u.r.pRule->pConstraints ;
-			ok && pConstraint ;
-			pConstraint = pConstraint->pNext )
-			{
-			ok = applyPATRConstraint(pDag, pConstraint, pData->pPATR);
-			}
+			ok = FALSE;
 		}
+		if (ok &&
+			(act_edge->eType == PATR_RULE_EDGE) &&
+			(act_edge->u.r.pRule->iNontermCount == act_edge->u.r.iNext))
+		{
+			ok = complete_constraints(pDag, act_edge->u.r.pRule, pData->pPATR);
 		}
+	}
 	/*
 	 *  allow the caller to kill us asynchronously, and to see our progress
 	 */
 	if (bCancelPATROperation_g || PumpMessages())
-		{
+	{
 		collectPATRGarbage(PATR_GARBAGE_UNIFY, pData->pPATR);
-		longjmp( sCancelPoint_m, 1 );
-		}
+		longjmp(sCancelPoint_m, 1);
+	}
 	++pData->uiProgressCount;
 	if ((pData->uiProgressCount % 10) == 0)
-		{
-		reportProgress( pData->uiProgressCount );
-		}
+	{
+		reportProgress(pData->uiProgressCount);
+	}
 	if (!ok && pData->pPATR->bUnification)
-		{
+	{
 		collectPATRGarbage(PATR_GARBAGE_UNIFY, pData->pPATR);
-		return;
-		}
-	unmarkPATRGarbage(PATR_GARBAGE_UNIFY, pData->pPATR);
+		*ok_out = ok;
+		*abort_out = TRUE;
+		return NULL;
 	}
+	unmarkPATRGarbage(PATR_GARBAGE_UNIFY, pData->pPATR);
 	/*
-	 *  eliminate edges whereby a single empty category becomes the
-	 *  daughter of two distinct non-terminals
+	 * Preserve rule dag for reunification.
 	 */
-	if (    (act_edge->iStart  == act_edge->iEnd)   &&
-		(act_edge->iStart  == iEnd_in)          &&
-		(pass_edge->iStart == pass_edge->iEnd)  &&
-		(pass_edge->iStart == act_edge->iStart) &&
-		(act_edge->eType   == PATR_RULE_EDGE)   &&
-		(pass_edge->eType  == PATR_RULE_EDGE) )
+	if (pDag)
 	{
-	/*
-	 *  find final nodes for both:  if same, don't add this edge
-	 */
-	if (get_final_empty_node(act_edge) == get_final_empty_node(pass_edge))
-		return;
-	}
+		if (act_edge->u.r.pChildren)
```

Comment:
should we check act_edge->eType here before assuming it's a rule?  maybe it's always the case by the time the code gets here...

Reply:
Lines 1272-1285 are no longer relevant.  They were part of an earlier design that involved reunification.  I will delete them.